### DISCUSSION TITLE
Is it possible to change data from outside?

### DISCUSSION THREAD

**Discussion by DahisC** [alpinejs/alpine#1374] [1. Help] [answered]
I am develop a website which is using Laravel blade engine (SSR) and Alpine to update views.

below is the sample of my code:

```html
<div x-data="alpineObj">
  // here will do foreach loop to render some cards
</div>
<div id="map">
  // will dynamic add some element into this element (using Leaflet map plugin)
  <div id="marker">will created by JavaScript</div>
</div>
```

```javascript
const alpineObj = alpine();
function alpine() {
  arr: [];
  updateArr(newArr) {
    this.arr = newArr;
  }
}
```

The situation is
I want to update binding data from outside.

I was trying to update alpine's binding data via `alpineObj` instance (like `alpineObj.updateArr([1,2,3])`), but when I changed that key `arr`, it didn't rerender the view, but `arr` has been changed.

is there have other way to do this or keywords? thanks a lot <3

**Comment by MuzafferDede** [comment, marked as answer]
you have few issues here,

- your object binding and function definition order is not correct
- you have some syntax errors `arr: [];`
- you should return object in your alpineObj function

[here](https://codepen.io/muzafferdede/pen/gOmdXjE) is a working codepen

**Reply by DahisC** [reply]
hello, thanks for your reply <3

sorry I forgot that return when I was writing this post, you're right.

and I tried some code on your codepen, like
```javascript
// change arr
alpineObj.arr = [1,2,3,4]
setTimeout(() => {
  alpineObj.updateArr([2,3,4,5]); // trying to use the method to update value of `arr`
  console.log(alpineObj.arr); // value is [2, 3, 4, 5], but view didn't change (it still showing [1, 2, 3 ,4])
}, 3000);
```

did I do something wrong?

**Reply by MuzafferDede** [reply]
Oh, you want to change the value from outside of alpine component, got you... In that case as i know you have to access alpine data proxy and call the updateArr function from there as:

```js
setTimeout(()=> {
  alpineObj.$el.__x.$data.updateArr([4,3,2,1])
},3000)
```
I have updated the [codepen](https://codepen.io/muzafferdede/pen/gOmdXjE?editors=1010) so you can have a look.


BTW, you can have more info about why and how here => [Access Alpine.js component data/state from outside its scope](https://codewithhugo.com/alpinejs-inspect-component-data-from-js/)

**Reply by DahisC** [reply]
yes, and I think that's the answer that I'am looking for!
it's very helpful, thank you :)

**Reply by arcanemachine** [reply]
> Oh, you want to change the value from outside of alpine component, got you... In that case as i know you have to access alpine data proxy and call the updateArr function from there as:
> 
> ```js
> setTimeout(()=> {
>   alpineObj.$el.__x.$data.updateArr([4,3,2,1])
> },3000)
> ```
> 
> I have updated the [codepen](https://codepen.io/muzafferdede/pen/gOmdXjE?editors=1010) so you can have a look.
> 
> BTW, you can have more info about why and how here => [Access Alpine.js component data/state from outside its scope](https://codewithhugo.com/alpinejs-inspect-component-data-from-js/)

The blog post link you provided doesn't work for Alpine v3.

I was able to get it to work by assigning a variable to the value of the selector containing the component:

`let el = document.querySelector('[x-data]');`

Then, I could access the data in `el._x_dataStack[0]`.

Of course, the devtools extension is a better way of doing all this, but it wasn't working for a while so I had to resort to hacks like this.
